File: methods/zeros/c-secant/secant.py
import math

def secant(f, xo, xa, n):
    for _ in range(n):
        xb = root(xo, xa)
        yield xb
        xo = xa
        xa = xb

def root(xo, xa):
    return (xo * f(xa) - xa * f(xo)) / (f(xa) - f(xo))

def f(x): 
    return 1161920 * math.exp(x) + (109945 / x) * (math.exp(x) - 1) - 2183290

iterations = [1, 2, 5]
xo = 0.10
xa = 1.73
n = 5

results = list(secant(f, xo, xa, n))
selection = []

for i in iterations:
    selection.append(results[i - 1])

print(*selection, sep = ",\n")

File: babireski/methods/zeros/c-secant/secant.c
#include <stdio.h>
#include <math.h>

double root(double (*f)(double), double xo, double xa)
{
    return (xo * f(xa) - xa * f(xo)) / (f(xa) - f(xo));
}

void secant(double (*f)(double), double xo, double xa, int n)
{
    int iterations[] = {1, 2, 5};
    int k = 0;

    for(int i = 0; i < n; i++)
    {
        if(f(xo) == f(xa))
        {
            return; 
        }

        double xb = root(f, xo, xa);

        if(iterations[k] == i + 1)
        {
            if(k)
            {
                printf(",\n");
            }

            printf("%.16f", xb);
            k++;
        }

        xo = xa;
        xa = xb;
    }
}

double f(double x)
{
    return sqrt(2 * 9.81 * x) * tanh(sqrt(2 * 9.81 * x) / (2 * 8.85) * 7.83) - 10.03;
}

int main()
{
    double xo = 0.6;
    double xa = 18.02;
    int n = 5;

    secant(f, xo, xa, n);

    return 0;
}